### ISSUE TITLE
BLAS was not found while installing other packages

### ISSUE THREAD

**Issue by xygao96** [xianyi/OpenBLAS#2205]
Hi,

Recently, I was trying to install some packages on clusters which requires BLAS/LAPACK. As the clusters are owned by the university, I do not have the privilege to use sudo to install libblas-dev and liblapack-dev. So I installed the OpenBLAS complied from the source files. However when I installed other packages (meep and scuff-em), I met some problems as the BLAS was not found.

What I did was first write:

export PATH=/home/gao477/OpenBLAS/bin:$PATH
export LD_LIBRARY_PATH=/home/gao477/OpenBLAS/lib:$LD_LIBRARY_PATH
export LIBRARY_PATH=/home/gao477/OpenBLAS/lib:$LIBRARY_PATH
export C_INCLUDE_PATH=/home/gao477/OpenBLAS/include:$C_INCLUDE_PATH

to the ./bashrc file to set the environment, after installing the OpenBLAS packages. 
And then, when I use sh autogen.sh to install other packages, I used the option:

sh autogen.sh --prefix=/home/gao477/packages/scuff-em -v LIBS=-ldl LDFLAGS="-L/home/gao477/Downloads/hdf5-1.10.5/hdf5/lib -L/home/gao477/OpenBLAS/lib  -L/usr/local/lib" CPPFLAGS="-I/usr/local/include -I/home/gao477/Downloads/hdf5-1.10.5/hdf5/include -I/home/gao477/OpenBLAS/include" 

to set path of OpenBLAS packages. However, when I did 'make', it still showed the error that 'OpenBLAS was not found'. The details are as follow:

checking for PTHREAD_PRIO_INHERIT... yes
checking hdf5.h usability... yes
checking hdf5.h presence... yes
checking for hdf5.h... yes
checking for H5Fcreate in -lhdf5... yes
checking for H5LTmake_dataset_double in -lhdf5_hl... yes
checking for sgemm_... no
checking for ATL_xerbla in -latlas... no
checking for sgemm_ in -lblas... no
checking for sgemm_ in -lmkl... no
checking for sgemm_... (cached) no
checking for sgemm_ in -lcxml... no
checking for sgemm_ in -ldxml... no
checking for sgemm_ in -lscs... no
checking for sgemm_ in -lcomplib.sgimath... no
checking for sgemm_ in -lblas... (cached) no
checking for sgemm_ in -lblas... (cached) no
configure: error: BLAS was not found!

From I found that hdf5 can be found but BLAS was not. And I didn't know what was wrong with my setting. Could you please help me figure it out? Thank you so much!

Xingyu

**Comment by isuruf** [comment]
> However, when I did 'make', it still showed the error that 'OpenBLAS was not found'. 

If you look at the error message carefully, it says `BLAS was not found`. It doesn't look for openblas at all. You should talk with the downstream projects about openblas support. (Or make a symlink `libblas.so -> libopenblas.so`)

**Comment by martin-frbg** [comment]
From a quick look at the scuff-em sources (in particular their m4/acx_blas.m4 that has the detection code), try `export BLAS_LIBS=-lopenblas` in addition to what you already exported.

**Comment by martin-frbg** [comment]
Same solution should work for meep (they use the same m4/acx_blas.m4 macro to detect BLAS). Also it should be fairly trivial to add OpenBLAS detection there by just copying their test for MKL
https://github.com/NanoComp/meep/blob/415bc8eb23c7578004e7a6ab5b44bf2a2bff21f3/m4/acx_blas.m4#L96-L99 and replacing both "mkl" on the AC_CHECK_LIB line with "openblas"

**Comment by xygao96** [comment]
Thank you for your replies! I just started to use linux last week, so I am not sure that I clearly understand all what you said. Sorry for that.
First, I tried to use:
ln -s libopenblas.so libblas.so
ln -s libopenblas.a libblas.a
to create a symlink and sh again. But this method did not work.

Then I write 'export BLAS_LIBS=-lopenblas' to ~/.bashrc and use 'source ~/.bashrc' to set the command line source. (This CentOS system doesn't have ~/.bashrc originally and I cannot change the etc file as I does't the privilige, so I need to set it manually.)

And for the last method you mentioned, I am not quite understand. Do you mean I need to add 

if test $acx_blas_ok = no; then
	AC_CHECK_LIB(openblas, $sgemm, [acx_blas_ok=yes;BLAS_LIBS="-lopenblas"])
fi

to acx_blas.m4 file to make the detection of openblas work?

Thank you again,
Xingyu

**Comment by martin-frbg** [comment]
Yes, that should work. (The test macro is run by "configure" when you execute the "autogen.sh" - basically it writes a small source file with a call to "sgemm" in it, and tries to compile and link that with 
each of the libraries it knows until it finds one that works.)

**Comment by xygao96** [comment]
Thank you. I tried again but still doesn't work. It is possible that I did not install OpenBLAS correctly? I remember I just used make and make PREFIX=/... install. It seems that there should be no problem with the installation. 

And I also tried to link the shared library with: 
gcc -o test test.c -I /home/gao477/OpenBLAS/include -L/home/gao477/OpenBLAS/lib -lopenblas
it gave an error:

/home/gao477/OpenBLAS/lib/libopenblas.so: file not recognized: File truncated
/usr/bin/sha1sum: test: No such file or directory
collect2: error: ld returned 1 exit status

And if I linked to static library with:
gcc -o test test.c /home/gao477/OpenBLAS/lib/libopenblas.a

it will also gave:
/home/gao477/OpenBLAS/lib/libopenblas.a(memory.o): In function `openblas_fork_handler':
memory.c:(.text+0x6ad): undefined reference to `pthread_atfork'
/home/gao477/OpenBLAS/lib/libopenblas.a(blas_server.o): In function `goto_set_num_threads':
blas_server.c:(.text+0x3ef): undefined reference to `pthread_create'
/home/gao477/OpenBLAS/lib/libopenblas.a(blas_server.o): In function `blas_thread_shutdown_':
blas_server.c:(.text+0xa77): undefined reference to `pthread_join'
/home/gao477/OpenBLAS/lib/libopenblas.a(blas_server.o): In function `blas_thread_init':
blas_server.c:(.text+0xc05): undefined reference to `pthread_create'
/home/gao477/OpenBLAS/lib/libopenblas.a(dgemm_oncopy.o): In function `dgemm_oncopy':
dgemm_ncopy_8_skylakex.c:(.text+0x7a1): undefined reference to `_intel_fast_memcpy'
/home/gao477/OpenBLAS/lib/libopenblas.a(dgemm_beta.o): In function `dgemm_beta':
dgemm_beta_skylakex.c:(.text+0x35a): undefined reference to `_intel_fast_memset'
dgemm_beta_skylakex.c:(.text+0x3d7): undefined reference to `_intel_fast_memset'
/bin/ld: link errors found, deleting executable `test'
/usr/bin/sha1sum: test: No such file or directory
collect2: error: ld returned 1 exit status

**Comment by martin-frbg** [comment]
Yes there does seem to be something wrong with your libopenblas.so ("file truncated" suggests it may be empty or very small). For linking with the static library, you will also need to add `-lpthread` for all the pthread functions. The two _intel_ functions are a bit strange however, did you use the Intel compiler (icc) to compile OpenBLAS ?

**Comment by xygao96** [comment]
> Yes there does seem to be something wrong with your libopenblas.so ("file truncated" suggests it may be empty or very small). For linking with the static library, you will also need to add `-lpthread` for all the pthread functions. The two _intel_ functions are a bit strange however, did you use the Intel compiler (icc) to compile OpenBLAS ?

Thank you so much! After re-installing the OpenBLAS and the method you metioned above works.
 
And the first problem (link the shared library) was fixed after re-install the package. But the second problem (static library) still exist. And I checked the make progress, it was compiled with mpiicc, which is shown as below:

OpenBLAS build complete. (BLAS CBLAS LAPACK LAPACKE)

  OS               ... Linux             
  Architecture     ... x86_64               
  BINARY           ... 64bit                 
  C compiler       ... INTEL  (command line : mpiicc)
  Fortran compiler ... INTEL  (command line : mpiifort)
  Library Name     ... libopenblas_haswellp-r0.3.7.dev.a (Multi threaded; Max num-threads is 40)

But anyway the main problem I met is well resolved. Thank you again! I feel I have bother you a lot.

Best,
Xingyu
